src/QsCompiler/QirGeneration/ScopeManager.cs

```diff
@@ -126,28 +126,106 @@ private static IEnumerable<(IValue, bool)> Expand(Func<ITypeRef, string?> getFun
             /// Increases the alias and reference count for the value if necessary,
-            /// and ensures that both are decreased again when the variable goes out of scope.
+            /// depending on whether the value is accessed via a mutable variable or assigned to a mutable variable.
+            /// Ensures that both are decreased again when the variable goes out of scope.
             /// The registered variable is mutable if the passed value is a pointer value.
             /// </summary>
-            public void RegisterVariable(string varName, IValue value)
+            public void RegisterVariable(string varName, IValue value, IValue? fromLocalId)
             {
                 this.variables.Add(varName, value);
 
-                // Since the value is necessarily created in the current or a parent scope,
-                // it won't go out of scope before the variable does.
-                // There is hence no need to increase the reference count if the variable is never rebound.
                 var change = new[] { (value, true) }; // true refers to whether the change also applies to inner items
                 this.increaseCounts(AliasCountUpdateFunctionForType, change);
+
+                // Since the value is necessarily created in the current or a parent scope,
+                // it won't go out of scope before the variable does.
+                // There is hence no need to increase the reference count unless either
+                // a) the value is assigned to a mutable variable (destination variable) that is later rebound, or
+                // b) the assigned value is accessed via a mutable variable (source variable) that is
+                //    rebound before the newly created variable goes out of scope.
+
+                // If either the source or destination variable can be rebound, however, then updating an item via a
+                // copy-and-reassign statement potentially leads to the updated item(s) being unreferenced in an inner scope,
+                // i.e. before the pending reference count increases of this scope are applied. We hence need to make sure
+                // to increase the reference count immediately when
+                // a) binding a value accessed via a local variable to a mutable variable, or
+                // b) assigning a value accessed via a mutable variable.
                 if (value is PointerValue)
                 {
-                    // If the variable can be rebound, however, then updating an item via a copy-and-reassign statement
-                    // potentially leads to the updated item(s) being unreferenced in an inner scope, i.e. before the
-                    // pending reference count increases of this scope are applied.
-                    // We hence need to make sure to increase the reference count immediately when binding to a mutable variable.
+                    // Assignment to a mutable variable:
```

Comment:
It seems that being a `PointerValue` is equivalent to being a mutable variable. I think it would make it easier to understand how the comments relate to the code by changing the code to `if(IsMutable(value))` as this makes the meaning of the statement explicit or start by explaining that we consider a variable mutable if it is an instance of `PointerValue`. Note it is non-obvious that PointerValue is equivalent to mutability as it could be a pointer to a const value.

Reply:
Hm, the explanation is largely valid even considering pointers more broadly, or put differently: I believe if this logic was invoked whenever a pointer is updated/defined it would be correct (albeit less efficient due to unnecessary alias count changes). The only reason we have special handling here is that we do some trickery with in-place modifications, and this logic is hence implementing optimizations specifically knowing which pointer values may be accessed in the source code later on. However, it would be fair to store other pointers in the scope manager, too; if that is done then we simply have some unnecessary overhead for those. 
Across other pieces of the code we do have PointerValues besides for things stored in mutable variables, and I hence believe that adding a function in the scope manager that equates a PointerValue with it being a mutable variable would be misleading.